include/seqan3/search/kmer_index/detail/shape_iterator.hpp

```diff
@@ -0,0 +1,461 @@
+ * text and a seqan3::shape that defines how to hash the text.
+ *
+ * \include test/snippet/search/kmer_index/shape_iterator_basic.cpp
+ *
+ * Depending on the type of the iterator passed to the shape_iterator, different functionality is available:
+ *
+ * | Concept modelled by passed text iterator                     | Available functions             |
+ * |--------------------------------------------------------------|---------------------------------|
+ * | \ref std::ForwardIterator "std::ForwardIterator"             | \ref shape_iterator_comparison "Comparison operators"<br>\ref operator++ "Pre-increment (++it)"<br>\ref operator++(int) "Post-increment (it++)"<br>\ref operator* "Indirection operator (*it)" |
+ * | \ref std::BidirectionalIterator "std::BidirectionalIterator" | \ref operator-- "Pre-decrement (--it)"<br>\ref operator--(int) "Post-decrement (it--)" |
+ * | \ref std::RandomAccessIterator "std::RandomAccessIterator"   | \ref operator+= "Forward (it +=)"<br>\ref operator+ "Forward copy (it +)"<br>\ref operator-= "Decrement(it -=)"<br>\ref shape_iterator_operator-decrement "Decrement copy (it -)"<br>\ref shape_iterator_operator-difference "Difference (it1 - it2)"<br>\ref operator[] "Subscript (it[])" |
+ *
+ * When using a gapped seqan3::shape, the `0`s of the seqan3::shape are virtually removed from the hashed k-mer.
+ * Note that any shape is expected to start with a `1` and end with a `1`.
+ *
+ * \snippet test/snippet/search/kmer_index/shape_iterator.cpp gapped
+ *
+ * The passed iterator will be treated as start position of the text, i.e. operator[] cannot be used to access a
+ * positions that precedes the start position.
+ *
+ * ### Implementation detail
+ *
+ * To avoid dereferencing the sentinel when iterating, e.g. when using a for loop:
+ * \snippet test/snippet/search/kmer_index/shape_iterator.cpp detail
+ * the shape_iterator computes the hash value up until the second to last position and performs the addition of the 
+ * last position upon access (\ref operator* and \ref operator[]).
+ */
+template <std::ranges::ForwardRange urng_t>
+    //!\cond
+    requires Semialphabet<delete_const_t<reference_t<urng_t>>>
+    //!\endcond
+class shape_iterator
+{
+private:
+    //!\brief The iterator type of the underlying range.
+    using it_t = std::ranges::iterator_t<urng_t>;
+    //!\brief The sentinel type of the underlying range.
+    using sentinel_t = std::ranges::sentinel_t<urng_t>;
+
+public:
+    /*!\name Associated types
+     * \{
+     */
+    //!\brief Type for distances between iterators.
+    using difference_type = typename it_t::difference_type;
+    //!\brief Value type of this iterator.
+    using value_type = size_t;
+    //!\brief Pointer to `value_type`.
+    using pointer = value_type *;
+    //!\brief Reference to `value_type`.
+    using reference = value_type &;
+    //!\brief Reference to `value_type const`.
+    using const_reference = value_type const &;
```

Comment:
Please double check these, I don't think they are correct, e.g. the reference type can't be `value_type &` as you are never returning references! Also the `pointer` type implies that you have `operator->` which you don't so I would remove it...

Reply:
I remember that something didn't work when I didn't have `pointer`, but now it works without.
It also seems that I do not need to define `reference` and `const_reference` - so should I remove it?

Reply:
I need the `pointer`:

```
/Users/enricoseiler/git/seqan3/include/seqan3/core/metafunction/iterator.hpp: In instantiation of 'struct seqan3::reference<seqan3::detail::shape_iterator<ranges::ref_view<std::vector<seqan3::dna4> > > >':
/Users/enricoseiler/git/seqan3/include/seqan3/core/metafunction/pre.hpp:71:7:   required by substitution of 'template<class t> using reference_t = typename seqan3::reference::type [with t = seqan3::detail::shape_iterator<ranges::ref_view<std::vector<seqan3::dna4> > >]'
/Users/enricoseiler/git/seqan3/include/seqan3/core/metafunction/range.hpp:78:11:   required from 'struct seqan3::reference<seqan3::detail::kmer_hash_view<ranges::ref_view<std::vector<seqan3::dna4> > > >'
/Users/enricoseiler/git/seqan3/include/seqan3/core/debug_stream.hpp:370:20:   required from here
/Users/enricoseiler/git/seqan3/include/seqan3/core/metafunction/iterator.hpp:56:11: error: no type named 'reference' in 'struct std::iterator_traits<seqan3::detail::shape_iterator<ranges::ref_view<std::vector<seqan3::dna4> > > >'
   56 |     using type = typename std::iterator_traits<std::remove_reference_t<it_t>>::reference;
      |           ^~~~
```

Reply:
You need reference, but you don't need pointer and const_reference. If something breaks without pointer being defined, set it to void.